[CI] ClusterInfoServiceIT testMaxQueueLatenciesInClusterInfo failing

**Build Scans:**
- [elasticsearch-periodic-platform-support #10025 / windows-2022_checkpart1_platform-support-windows](https://gradle-enterprise.elastic.co/s/2rjuh6edqvxb4)
- [elasticsearch-pull-request #86036 / part-1](https://gradle-enterprise.elastic.co/s/vmkbzfqefrhli)
- [elasticsearch-periodic-platform-support #10025 / almalinux-9_platform-support-unix](https://gradle-enterprise.elastic.co/s/lrmdnd7du6oz2)
- [elasticsearch-periodic-platform-support #10025 / debian-12_platform-support-unix](https://gradle-enterprise.elastic.co/s/czmor4yywbvvs)
- [elasticsearch-pull-request #86008 / part-1](https://gradle-enterprise.elastic.co/s/7u24mc3p3iqgy)
- [elasticsearch-pull-request #86000 / part-1](https://gradle-enterprise.elastic.co/s/k2f4sswu67nja)

**Reproduction Line:**
```
gradlew ":server:internalClusterTest" --tests "org.elasticsearch.cluster.ClusterInfoServiceIT.testMaxQueueLatenciesInClusterInfo" -Dtests.seed=F5440B5D5A6B1B4C -Dtests.locale=kxv-Orya -Dtests.timezone=SystemV/MST7 -Druntime.java=24
```

**Applicable branches:**
main

**Reproduces locally?:**
N/A

**Failure History:**
[See dashboard](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupState:(initialChildControlState:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:task.keyword,order:0,selectedOptions:!(),title:'GradleTask',type:optionsListControl),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:className.keyword,order:1,selectedOptions:!(org.elasticsearch.cluster.ClusterInfoServiceIT),title:'Suite',type:optionsListControl),'144933da-5c1b-4257-a969-7f43455a7901':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:name.keyword,order:2,selectedOptions:!('testMaxQueueLatenciesInClusterInfo'),title:'Test',type:optionsListControl)))))

**Failure Message:**
```
java.lang.AssertionError: Unexpectedly found a task queued for the write thread pool. Write thread pool dump: TaskExecutionTimeTrackingEsThreadPoolExecutor[name = node_t1/write, queue capacity = 10000, task execution EWMA = 17.7ms, total task execution time = 1s, org.elasticsearch.common.util.concurrent.TaskExecutionTimeTrackingEsThreadPoolExecutor@12a864d[Running, pool size = 1, active threads = 1, queued tasks = 3, completed tasks = 10]]
Expected: <0L>
     but: was <1L>
```

**Issue Reasons:**
- [main] 6 failures in test testMaxQueueLatenciesInClusterInfo (13.3% fail rate in 45 executions)
- [main] 3 failures in step part-1 (11.5% fail rate in 26 executions)
- [main] 3 failures in pipeline elasticsearch-pull-request (12.0% fail rate in 25 executions)

**Note:**
This issue was created using new test triage automation. Please report issues or feedback to es-delivery.

--- a/muted-tests.yml
+++ b/muted-tests.yml
@@ -462,9 +462,6 @@ tests:
 - class: org.elasticsearch.search.CCSDuelIT
   method: testTermsAggsWithProfile
   issue: https://github.com/elastic/elasticsearch/issues/132880
-- class: org.elasticsearch.cluster.ClusterInfoServiceIT
-  method: testMaxQueueLatenciesInClusterInfo
-  issue: https://github.com/elastic/elasticsearch/issues/132957
 - class: org.elasticsearch.xpack.search.AsyncSearchErrorTraceIT
   method: testAsyncSearchFailingQueryErrorTraceDefault
   issue: https://github.com/elastic/elasticsearch/issues/133010

--- a/server/src/internalClusterTest/java/org/elasticsearch/cluster/ClusterInfoServiceIT.java
+++ b/server/src/internalClusterTest/java/org/elasticsearch/cluster/ClusterInfoServiceIT.java
@@ -27,9 +27,11 @@ import org.elasticsearch.cluster.routing.allocation.decider.EnableAllocationDeci
 import org.elasticsearch.cluster.service.ClusterService;
 import org.elasticsearch.common.Strings;
 import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.util.concurrent.EsExecutors;
 import org.elasticsearch.common.util.concurrent.TaskExecutionTimeTrackingEsThreadPoolExecutor;
 import org.elasticsearch.core.TimeValue;
 import org.elasticsearch.index.IndexService;
+import org.elasticsearch.index.shard.GlobalCheckpointListeners;
 import org.elasticsearch.index.shard.IndexShard;
 import org.elasticsearch.index.shard.ShardId;
 import org.elasticsearch.index.store.Store;
@@ -55,6 +57,7 @@ import java.util.Locale;
 import java.util.Map;
 import java.util.Set;
 import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.Executor;
 import java.util.concurrent.atomic.AtomicBoolean;
 
 import static java.util.Collections.emptySet;
@@ -423,9 +426,12 @@ public class ClusterInfoServiceIT extends ESIntegTestCase {
         var masterName = internalCluster().startMasterOnlyNode(settings);
         var dataNodeName = internalCluster().startDataOnlyNode(settings);
         ensureStableCluster(2);
-        assertEquals(internalCluster().getMasterName(), masterName);
-        assertNotEquals(internalCluster().getMasterName(), dataNodeName);
-        logger.info("---> master node: " + masterName + ", data node: " + dataNodeName);
+
+        String indexName = randomIdentifier();
+        final int numShards = randomIntBetween(1, 5);
+        createIndex(indexName, Settings.builder().put(SETTING_NUMBER_OF_SHARDS, numShards).put(SETTING_NUMBER_OF_REPLICAS, 0).build());
+        ensureGreen(indexName);
+        final var indexService = internalCluster().getInstance(IndicesService.class, dataNodeName).iterator().next();
 
         // Block indexing on the data node by submitting write thread pool tasks equal to the number of write threads.
         var barrier = blockDataNodeIndexing(dataNodeName);
@@ -434,12 +440,6 @@ public class ClusterInfoServiceIT extends ESIntegTestCase {
             // strictly need a single task to occupy the queue.
             int numberOfTasks = randomIntBetween(1, 5);
             Thread[] threadsToJoin = new Thread[numberOfTasks];
-            String indexName = randomIdentifier();
-            createIndex(
-                indexName,
-                // NB: Set 0 replicas so that there aren't any stray GlobalCheckpointSyncAction tasks on the write thread pool.
-                Settings.builder().put(SETTING_NUMBER_OF_SHARDS, randomIntBetween(1, 5)).put(SETTING_NUMBER_OF_REPLICAS, 0).build()
-            );
             for (int i = 0; i < numberOfTasks; ++i) {
                 threadsToJoin[i] = startParallelSingleWrite(indexName);
             }
@@ -494,6 +494,30 @@ public class ClusterInfoServiceIT extends ESIntegTestCase {
             for (int i = 0; i < numberOfTasks; ++i) {
                 threadsToJoin[i].join();
             }
+            Arrays.stream(threadsToJoin).forEach(thread -> assertFalse(thread.isAlive()));
+
+            // Wait for async post replication actions to complete
+            final var checkpointsSyncLatch = new CountDownLatch(numShards);
+            for (int i = 0; i < numShards; ++i) {
+                final var indexShard = indexService.getShard(i);
+                final long expectedGlobalCheckpoint = indexShard.seqNoStats().getGlobalCheckpoint();
+                logger.info("--> shard [{}] waiting for global checkpoint {}", i, expectedGlobalCheckpoint);
+                indexShard.addGlobalCheckpointListener(expectedGlobalCheckpoint, new GlobalCheckpointListeners.GlobalCheckpointListener() {
+                    @Override
+                    public Executor executor() {
+                        return EsExecutors.DIRECT_EXECUTOR_SERVICE;
+                    }
+
+                    @Override
+                    public void accept(long globalCheckpoint, Exception e) {
+                        assertNull(e); // should have no error
+                        logger.info("--> shard [{}] global checkpoint updated to {}", indexShard.shardId().id(), globalCheckpoint);
+                        checkpointsSyncLatch.countDown();
+                    }
+                }, TimeValue.THIRTY_SECONDS);
+            }
+            safeAwait(checkpointsSyncLatch);
+
             assertThat(
                 "Unexpectedly found a task queued for the write thread pool. Write thread pool dump: " + trackingWriteExecutor,
                 trackingWriteExecutor.peekMaxQueueLatencyInQueueMillis(),

